File: hadesboot/src/main/kotlin/hadesc/location/SourceLocation.java
package hadesc.location;

public record SourceLocation(
        SourcePath file,
        Position start,
        Position stop
) implements Comparable<SourceLocation>, HasLocation {

    @Override
    public String toString() {
        return "($file:${start.line})";
    }

    @Override
    public SourceLocation getLocation() {
        return this;
    }

    public boolean isWithin(SourceLocation other) {
        return (start.gte(other.start)) && (stop.lte(other.stop));
    }

    public boolean contains(HasLocation node) {
        return node.getLocation().isWithin(this);
    }

    public static SourceLocation between(HasLocation start, HasLocation stop) {
        return new SourceLocation(
                start.getLocation().file,
                start.getLocation().start,
                stop.getLocation().stop
        );
    }

    @Override
    public int compareTo(SourceLocation other) {
        if (start.equals(other.start)) {
            if (stop.lte(other.stop)) {
                return 1;
            } else {
                return -1;
            }

        } else if (start.gte(other.start)) {
            return 1;
        } else {
            return -1;
        }
    }

}

File: hadesboot/src/main/kotlin/hadesc/location/SourceLocation.kt
package hadesc.location

data class SourceLocation(
    val file: SourcePath,
    val start: Position,
    val stop: Position,
) : Comparable<SourceLocation>, HasLocation {
    override fun toString(): String {
        return "(" + file + ":" + start.line() + ")"
    }

    override val location: SourceLocation
        get() = this

    fun isWithin(other: SourceLocation): Boolean {
        return start.gte(other.start) && stop.lte(other.stop)
    }

    operator fun contains(node: HasLocation): Boolean {
        return node.location.isWithin(this)
    }

    override fun compareTo(other: SourceLocation): Int {
        return if (start == other.start) {
            if (stop.lte(other.stop)) {
                1
            } else {
                -1
            }
        } else if (start.gte(other.start)) {
            1
        } else {
            -1
        }
    }

    companion object {
        fun between(start: HasLocation, stop: HasLocation): SourceLocation {
            return SourceLocation(
                start.location.file,
                start.location.start,
                stop.location.stop
            )
        }
    }
}